server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -531,27 +561,30 @@ public RetryOnReplicaException(StreamInput in) throws IOException {
             try {
-                assert replica.getActiveOperationsCount() != 0 : "must perform shard operation under a permit";
-                final ReplicaResult replicaResult = shardOperationOnReplica(replicaRequest.getRequest(), replica);
-                replicaResult.runPostReplicaActions(
-                    ActionListener.wrap(r -> {
-                        final TransportReplicationAction.ReplicaResponse response =
-                            new ReplicaResponse(replica.getLocalCheckpoint(), replica.getLastSyncedGlobalCheckpoint());
-                        releasable.close(); // release shard operation lock before responding to caller
-                        if (logger.isTraceEnabled()) {
-                            logger.trace("action [{}] completed on shard [{}] for request [{}]", transportReplicaAction,
-                                replicaRequest.getRequest().shardId(),
-                                replicaRequest.getRequest());
-                        }
-                        setPhase(task, "finished");
-                        onCompletionListener.onResponse(response);
-                    }, e -> {
-                        Releasables.closeWhileHandlingException(releasable); // release shard operation lock before responding to caller
-                        this.responseWithFailure(e);
-                    })
-                );
-            } catch (final Exception e) {
+                shardOperationOnReplica(replicaRequest.getRequest(), replica, ActionListener.wrap((replicaResult) ->
+                    replicaResult.runPostReplicaActions(
+                        ActionListener.wrap(r -> {
+                            final ReplicaResponse response =
+                                new ReplicaResponse(replica.getLocalCheckpoint(), replica.getLastSyncedGlobalCheckpoint());
+                            releasable.close(); // release shard operation lock before responding to caller
+                            if (logger.isTraceEnabled()) {
+                                logger.trace("action [{}] completed on shard [{}] for request [{}]", transportReplicaAction,
+                                    replicaRequest.getRequest().shardId(),
+                                    replicaRequest.getRequest());
+                            }
+                            setPhase(task, "finished");
+                            onCompletionListener.onResponse(response);
+                        }, e -> {
+                            Releasables.closeWhileHandlingException(releasable); // release shard operation lock before responding to caller
+                            responseWithFailure(e);
+                        })
+                    ), e -> {
+                    Releasables.closeWhileHandlingException(releasable); // release shard operation lock before responding to caller
+                    AsyncReplicaAction.this.onFailure(e);
+                }));
+            } catch (Exception e) {
```

Comment:
I think we can remove this catch block now? `shardOperationOnReplica` should never fail with an exception, only case is if the listener fails, which we are asserting against in `completeWith`.